File: FCARDIO.Core/FCARDIO.Core/Extension/ByteBufferExtension.vb
Imports System.Runtime.CompilerServices
Imports System.Runtime.InteropServices
Imports DotNetty.Buffers

Namespace Extension
    Public Module ByteBufferExtension
        <StructLayout(LayoutKind.Explicit)>
        Public Structure UnionInt32
            <FieldOffset(0)>
            Public IntValue As Int32
            <FieldOffset(0)>
            Public UIntValue As UInt32
        End Structure

        <StructLayout(LayoutKind.Explicit)>
        Public Structure UnionInt64
            <FieldOffset(0)>
            Public IntValue As Int64
            <FieldOffset(0)>
            Public UIntValue As UInt64
        End Structure


        ''' <summary>
        ''' 将UINT32类型写入到缓冲区
        ''' </summary>
        ''' <param name="buf"></param>
        ''' <param name="v"></param>
        ''' <returns></returns>
        <Extension()>
        Public Function WriteUnsignedInt(buf As IByteBuffer, v As UInt32) As IByteBuffer
            Dim t As UnionInt32
            t.UIntValue = v
            Return buf.WriteInt(t.IntValue)
        End Function

        ''' <summary>
        ''' 将UINT64类型写入到缓冲区
        ''' </summary>
        ''' <param name="buf"></param>
        ''' <param name="v"></param>
        ''' <returns></returns>
        <Extension()>
        Public Function WriteUInt64(buf As IByteBuffer, v As UInt64) As IByteBuffer
            Dim t As UnionInt64
            t.UIntValue = v
            Return buf.WriteLong(t.IntValue)
        End Function

        <Extension()>
        Public Function ToHex(buf As IByteBuffer) As String
            Return ByteBufferUtil.HexDump(buf)
        End Function
    End Module

End Namespace

File: tag/v1.x/DoNetDrive.Protocol.Util/DoNetDrive.Protocol.Util/ByteBufferExtension.cs
using System;
using System.Collections.Generic;
using System.Text;
using DotNetty.Buffers;
namespace DoNetDrive.Protocol.Util
{
    /// <summary>
    /// 缓存区扩展函数
    /// </summary>
    public static class ByteBufferExtension
    {
        /// <summary>
        /// 写入一个uint32数值到缓存区
        /// </summary>
        /// <param name="buf"></param>
        /// <param name="value"></param>
        /// <returns></returns>
        public static IByteBuffer WriteUInt32(this IByteBuffer buf, uint value)
        {
            return buf.WriteInt((int)value);
        }
        /// <summary>
        /// 写入一个uint64数值到缓存区
        /// </summary>
        /// <param name="buf"></param>
        /// <param name="value"></param>
        /// <returns></returns>
        public static IByteBuffer WriteUInt64(this IByteBuffer buf, ulong value)
        {
            return buf.WriteLong((long)value);
        }
        /// <summary>
        /// 将缓存中数据编码为HEX
        /// </summary>
        /// <param name="buf"></param>
        /// <returns></returns>
        public static string ToHex(this IByteBuffer buf)
        {
            return ByteBufferUtil.HexDump(buf);
        }
    }
}
